Update documentation and warnings before `0.1` release

rhugonnet opened the pull request:
This PR improves the documentation in preparation for a first minor release `0.1`! :partying_face: 

The documentation now separates scientific background content and code examples, adds better links and plots to all feature pages, and adds some auxiliary pages for diverse uses based on user feedback (how to cite, method overview, cheatsheet for helping selecting a method, configuration, ecosystem, etc).

Link to rendered doc: https://xdem-rhugonnet.readthedocs.io/en/towards_0.1/index.html 

## Detailed documentation changes

This PR:
- Moves scientific background content that was mixed with the code examples, and adds new content, into a series of **guides pages** "Guides for elevated analysis" in the new section "Resources", illustrated with plots generated just for the documentation and a list for references for "More reading" at the bottom of the pages,
- Adds a "Cheatsheet: How to correct... ?" page in the new section "Resources" to help users read patterns of errors in elevation data to select a correction method,
- Adds an "Ecosystem" page in the new section "Resources" that points towards other relevant packages to pair with xDEM,
- Adds a "Citing and method overview" page in the section "Getting started", that points to xDEM's Zenodo and lists the references to use for all methods implemented in the packages,
- Adds a "Use in publication" page in the section "Reference", that lists scientific articles/theses that are using xDEM (to our knowledge),
- Adds a "Configuration" page in the section "Reference", that explains how to configure verbosity level, and georeferenced operations in xDEM via GeoUtils (resampling, etc),
- Streamlines and adds plots to the "Coregistration", "Bias correction" and "Uncertainty analysis" **features pages** for better understanding of these methods,
- Add "Quick use" subsections at the start of all feature pages, to save time for user who don't want to get into the details,
- Adds tables listing all methods at the start of "Terrain attributes", "Coregistration" and "Uncertainty analysis",
- Adds up-to-date descriptions of modular arguments and metadata for "Coregistration" based on recent changes,
- Adds equations used for "Terrain attributes",
- Adds more links to GeoUtils with explanation of the inheritance between packages,
- Adds links between **feature pages** and **guide pages**,
- Streamlines the API reference into a manually ordered list of submodule functions meant to be publicly referenced, and repeating functions that are from GeoUtils to facilitate user searches.

## Other small changes

This PR also has some small changes in the core modules related to the documentation/visualization:
- Adds `AffineCoreg.to_translations()` and `AffineCoreg.to_rotations()` and `AffineCoreg.from_rotations()` (as `from_translations()` already existed) to quickly extract from/build the affine matrix, which might not be easy otherwise for a user (added in a note on the Coregistration page), which directly use `pytransform3d` (so don't require testing),
- Adds a `resample` option to `apply_matrix` for practicality (using a the subfunction `_reproject_horizontal_shift_samecrs` that was already tested, and is moved to the `base` module instead of `affine`),
- Adds more formatting options to `Coreg.info()` for types that were not yet supported (after some testing in the documentation), and a `CoregPipeline.info()` that simply wraps the `Coreg.info()` of each pipeline step,
- Updates the recent function `DEM.estimate_uncertainty()` to have clearer arguments, choice of method mirroring the new documentation.

## TO-DO:

- [x] Add profile curvature and plan curvature formulas on "Terrain attributes" page,
- [x] Add `BlockwiseCoreg` to coregistration page,
- [x] Add "Cheatsheet" page,
- [x] Add link to PyTransform3d on Affine registration page, 
- [x] Add "Ecosystem" page,
- [x] Add "Georeferencing intricacies" page,
- [x] Add documentation for "DhMinimize" method,
- [x] Streamline "Coregistration" and "Bias correction" pages, 
- [x] Detail `.meta` keys for users (and make the attribute public in a separate PR),
- [x] Add `.info()` method for `CoregPipeline`,
- [x] Remove temporarily `BlockwiseCoreg`,
- [x] Update reference for TPI: #562,
- [x] Re-structure/streamline "Spatial statistics for error analysis" guide page,
- [x] Finalize "Uncertainty analysis" page, add warning that interface might change with Scikit-GStat now supporting directly sum of variogram models,
- [x] Streamline "Differencing and volume change" into a "Gap-filling" page and add warning that it might be moved to a different package,
- [x] Proof-read/streamline all gallery examples,
- [x] Prepare banner to communicate breaking changes with `0.1`,
- [x] Check if figures in "Quick start" are fixed (normal colorbar, vector plotting on top),
- [x] Check if PROJ error still appears or not in the rendered documentation, fix it if yes,
- [x] Add "read more" section at the end of each guide page?
- [x] Move down `dDEM` and `DEMCollection` in API, remove `dDEM` from README, and add warning of potential removal in future,
- [x] Fix Nuth&Kääb gallery example colorbar title, 
- [x] Shorten detailed intros from gallery example, 
- [x] Fix all hyperlinks and warnings,
- [x] Fix warnings of sphinx build,
- [x] Release new GeoUtils with default tight layout,
- [ ] Fix cheatsheet vertical CRS failing to download on RTD, 
- [x] Fix bias-correction sinusoid example sometimes failing to correct randomly. Couldn't! Opened #637 

Resolves #477 (last step)
Resolves #505
Resolves #464
Resolves #434
Resolves #285
Resolves #275
Resolves #431
Resolves #532 
Resolves #528
Resolves #562 
Resolves #583
Resolves #614

Opens #637
Opens #638

erikmannerfelt commented on doc/source/accuracy_precision.md:
```diff
@@ -0,0 +1,115 @@
+
+### What are accuracy and precision?
+
+[Accuracy and precision](https://en.wikipedia.org/wiki/Accuracy_and_precision) describe random and systematic errors, respectively.
```
Flip "random and systematic", as accuracy is systematic and precision is random.

rhugonnet replied:
Good catch! Fixed

erikmannerfelt commented on doc/source/accuracy_precision.md:
```diff
@@ -0,0 +1,115 @@
+
+### Translating these concepts for elevation data
+
+However, elevation data rarely consists of a single independent measurement but of a **series of measurement** (image grid,
```
"series of measurement**s**"

rhugonnet replied:
Fixed

erikmannerfelt commented on doc/source/accuracy_precision.md:
```diff
@@ -0,0 +1,115 @@
+Accuracy is generally considered from two focus points:
+
+- **Absolute elevation accuracy** describes systematic errors to the true positioning, usually important when analysis focuses on the exact location of topographic features at a specific epoch.
+- **Relative elevation accuracy** describes systematic errors with reference to other elevation data that does not necessarily matches the true positioning, important for analyses interested in topographic change over time.
```
"does not necessarily **match** the true positioning"

erikmannerfelt commented on doc/source/conf.py:
```diff
@@ -25,8 +25,8 @@
 project = "xDEM"
-copyright = "2021, Erik Mannerfelt, Romain Hugonnet, Amaury Dehecq and others"
-author = "Erik Mannerfelt, Romain Hugonnet, Amaury Dehecq and others"
+copyright = "2020, GlacioHack"
```
I see some packages leaving the original year while others keep it updated. Any ideas on implications of this?

rhugonnet replied:
No, we should ask advice from CNES/CS on this. Maybe we can adjust when we'll change the license.

erikmannerfelt commented on doc/source/ecosystem.md:
```diff
@@ -0,0 +1,27 @@
+(ecosystem)=
+
+# Ecosystem
+
```
Maybe not here, but somewhere; should we refer to demcompare as that's still used quite much? Sorry if it already says somewhere.

rhugonnet replied:
Good point! Added for now, CNES/CS can orient us on how they'd like demcompare to show up in the doc.

erikmannerfelt commented on doc/source/ecosystem.md:
```diff
@@ -0,0 +1,27 @@
+# Ecosystem
+
+xDEM is but a single tool among a large landscape of open tools for geospatial elevation analysis! Below is a list of 
+other **tools that you might find useful to combine with xDEM**, in particular for retrieving elevation data or to perform complementary analysis.
```
Add geoutils, as it might not be evident how closely related these two projects are.

rhugonnet replied:
Good idea, done!

erikmannerfelt commented on doc/source/spatial_stats.md:
```diff
@@ -0,0 +1,186 @@
+For more details on variography, **we refer to [the documentation of SciKit-GStat](https://scikit-gstat.readthedocs.io/en/latest/userguide/userguide.html).**
+
+
+## Error propagation
```
Add a TODO here or remove it!

erikmannerfelt replied:
Or an issue

rhugonnet replied:
Was unfinished, now done!

erikmannerfelt commented on doc/source/static_surfaces.md:
```diff
@@ -0,0 +1,81 @@
+
+## The great benefactor of quantitative elevation analysis
+
+Elevation data benefits from an uncommon asset, which is that **large proportions of planetary surface elevations
+usually remain virtually unchanged through time**. Those static surfaces, sometimes also referred to as "stable terrain",
```
The geologist in me wants you to add "within human time-scales". Everything moves significantly if you wait long enough!

rhugonnet replied:
Good point, added!

erikmannerfelt commented on doc/source/static_surfaces.md:
```diff
@@ -0,0 +1,81 @@
+
+Elevation data benefits from an uncommon asset, which is that **large proportions of planetary surface elevations
+usually remain virtually unchanged through time**. Those static surfaces, sometimes also referred to as "stable terrain",
+generally refer to bare-rock, grasslands, and are often isolated by excluding dynamic surfaces such as glaciers,
```
..., rivers, human settlements, lakes, ....

erikmannerfelt commented on xdem/coreg/biascorr.py:
```diff
@@ -460,7 +460,7 @@ def __init__(
         | Literal["nfreq_sumsin"] = "norder_polynomial",
         fit_optimizer: Callable[..., tuple[NDArrayf, Any]] = scipy.optimize.curve_fit,
-        bin_sizes: int | dict[str, int | Iterable[float]] = 100,
+        bin_sizes: int | dict[str, int | Iterable[float]] = 1000,
```
Does this link to an issue? I'm sure it's a good update but it's a bit off topic!

rhugonnet replied:
Those default parameters were a bit arbitrary (one of my PRs 2 years ago) and didn't render well, noticed that in the doc so adjusted it!

erikmannerfelt commented on xdem/coreg/base.py:
```diff
@@ -2627,7 +2701,7 @@ def copy(self: CoregType) -> CoregType:
         new_coreg = self.__new__(type(self))
 
-        new_coreg.__dict__ = {key: copy.deepcopy(value) for key, value in self.__dict__.items() if key != "pipeline"}
+        new_coreg.__dict__ = {key: copy.copy(value) for key, value in self.__dict__.items() if key != "pipeline"}
```
What does this change do?

rhugonnet replied:
Forgot to add this in the `CoregPipeline` to mirror an issue fixed in #530 for `Coreg`

erikmannerfelt commented on xdem/coreg/affine.py:
```diff
@@ -848,6 +788,34 @@ def to_matrix(self) -> NDArrayf:
 
+    def to_translations(self) -> tuple[float, float, float]:
+        """
+        Convert the affine transformation matrix to only its X/Y/Z translations.
```
I think something funky will happen here if scaling is involved (I don't remember which order the components are applied in). I feel like a warning should be added to this and the `to_rotations()` saying something like "Warning: Information may be lost".

And the conversion is less of a conversion and more of an extraction I would say! Since there's no validation that scaling won't mess with the result.

rhugonnet replied:
I agree that it's not a conversion, the idea was just to have an extraction! We could change this in the description? 
Leaving this conversation open for @adehecq to comment.

rhugonnet replied:
And I'll look into it for when scaling is involved, great point :wink:

rhugonnet replied:
For the record: translation is OK with scaling, but the rotation is affected. Opening an issue to add this in the future.

erikmannerfelt commented on tests/test_coreg/test_base.py:
```diff
@@ -960,8 +960,7 @@ def test_blockwise_coreg_large_gaps(self) -> None:
         ddem_pre = (tba - self.ref).data.compressed()
         assert abs(np.nanmedian(ddem_pre)) > abs(np.nanmedian(ddem_post))
-        # TODO: Figure out why STD here is larger since PR #530
-        # assert np.nanstd(ddem_pre) > np.nanstd(ddem_post)
```
Why is an identical assertion commented out here?

rhugonnet replied:
This is linked to #584

erikmannerfelt reviewed (commented):
Wow, such an improvement to the documentation!! Super great job @rhugonnet!

I have some questions, generally pertaining to the non-documentation part of the PR (perhaps in the future changes like this could be a separate PR? :) )

rhugonnet commented on xdem/coreg/affine.py:
```diff
@@ -1521,7 +1520,7 @@ def _fit_rst_pts(
         # Write output to class
         # (Mypy does not pass with normal dict, requires "OutAffineDict" here for some reason...)
-        output_affine = OutAffineDict(shift_x=-easting_offset, shift_y=-northing_offset, shift_z=vertical_offset)
+        output_affine = OutAffineDict(shift_x=easting_offset, shift_y=northing_offset, shift_z=vertical_offset)
```
This line removes a temporary fix that was done in parallel of #595 to run this PR.
So there is actually no change at this line when comparing to `main`! (but you had it pop up when clicking on "See changes since last review")

rhugonnet commented:
@vschaffn @adebardo FYI: I used the opportunity of merging the new `main` in this big documentation PR to integrate the documentation on logging of #610 (because the page  was not listed in the index so not rendered, and same for the Sphinx-Gallery example as it had the wrong name pattern!).

If you want to check the rendering of the docs in advance in next PRs, you can add your fork (xdem-yourfork) in Readthedocs to render certain branches (Import Project, then Versions/Activate a version/Choose hidden, then the doc builds automatically in Builds at each commit on that branch!): https://readthedocs.org/dashboard/ :wink: 

In short, I have made the following small changes:
- Made both page and example render,
- Integrated the "logging" page into a section of a more generic "Configuration" page,
- Added the name "Verbosity level" in addition to "Logging" in titles and description as, after a bit of asking around, few people knew what logging referred to!
- Re-structured the Sphinx-Gallery example to only show logging within the xDEM function calls for clarity (otherwise it was a bit much at once, after seeing the rendered output!),
- Added a one-liner example for those who just want to know how to show "logging.INFO" level quickly,
- Added some Sphinx configuration-file specifics to make it work (resetting the logging between examples).

rhugonnet commented:
@adehecq @erikmannerfelt This PR is finalized, will be merging soon. Only two small issues I can't seem to fix (with `DirectionalBias` and vertical CRS download through readthedocs...), will likely open issues.
I'll then open a separate PR to add documentation notes on the fusion with demcompare, and get feedback from CNES/CS.

adehecq commented on doc/source/background.md:
```diff
@@ -0,0 +1,66 @@
+
+The initial core development of xDEM was performed by members of the Glaciology group of the Laboratory of Hydraulics, Hydrology and
+Glaciology (VAW) at ETH Zürich<sup>3</sup>, with contributions by members of the University of Oslo, the University of Washington, and University
+Grenobles Alpes.
```
Grenoble without s :wink:

adehecq commented on doc/source/quick_start.md:
````diff
@@ -46,6 +56,11 @@ fn_glacier_outlines = xdem.examples.get_path("longyearbyen_glacier_outlines")
+:class: margin
+Set up your {ref}`verbosity` to manage outputs to the console (or a file) during execution!
+```
+
````
Just a question on line 78 below, which was not edited. 
I assume this is just for the example, as both DEMs are originally on the same vertical datum? It must impact coregistration a bit... I think it's great to showcase this functionality, but it woul dbe good to have a sample data set with actual vertical datum difference.

adehecq replied:
+ L 119: "An basic" -> "a basic"

rhugonnet replied:
Good catch. And opened https://github.com/GlacioHack/xdem/issues/635 for the vertical datum of the DEM.

adehecq commented on doc/source/citation.md:
```diff
@@ -0,0 +1,114 @@
+
+## Citing xDEM
+
+To cite the package, use the Zenodo DOI: [![Zenodo](https://zenodo.org/badge/doi/10.5281/zenodo.4809697.svg)](https://zenodo.org/doi/10.5281/zenodo.4809697).
```
I think we need to be more detailed than this. The best way is to have a ref like "(xDEM contributors, 202X)" and in the reference list: "Xdem Contributors: xdem, Zenodo [code], https://doi.org/10.5281/zenodo.YYYY, 202X." with year and zenodo link to updated depending on the version and year used.

adehecq replied:
It will be easier when we have a proper article for it...

rhugonnet replied:
The link I put is the generic Zenodo link that works for all versions.
I wanted to add a comment about the option to point to a specific version, but we haven't done this so well until now... (We only have 0.0.2 and 0.0.18, which is a bit weird).

@erikmannerfelt Can you activate user sharing on the Zenodo for @adehecq and me (https://help.zenodo.org/docs/share/user-sharing/), so that any one of us can update the versioning when others are too busy. Or maybe we can even set-up auto-updates on version + authors? I've seen several packages do it.

rhugonnet replied:
My bad! I forgot you added us, and we can edit when logged in Zenodo with our ORCID. All good!

adehecq commented on doc/source/citation.md:
```diff
@@ -0,0 +1,114 @@
+     - [Girod et al. (2017)](https://doi.org/10.3390/rs9070704)
+   * - Directional bias (other)
+     - N/A
+   * - Terrain bias (maximum curvature)
```
Hopefully we can update this with the work of Arnaud!

rhugonnet replied:
Yes! :slightly_smiling_face:

rhugonnet replied:
I thought exactly the same when redacting it.

adehecq commented on doc/source/citation.md:
```diff
@@ -0,0 +1,114 @@
+   * - Bilinear
+     - N/A
+   * - Local and regional hypsometric
+     - [McNabb et al. (2019)](https://tc.copernicus.org/articles/13/895/2019/)
```
It is definitely a good resource to understand the concept of the hypsometric approach, but this is not the paper that actually proposes this method. It has been used for years before Bob's paper. I will check if I can find an original reference for this.

adehecq replied:
McNabb 2019 refers to Arendt (2002) and Kääb (2008). In the first one, there is no occurrence of "hypso". The second one uses the term "hypsography" ! So not sure what is the best way to acknowledge...

rhugonnet replied:
Good point. Indeed, Anthony is binning per elevation band in his study. I trust @iamdonovan will have done the work on this :stuck_out_tongue:  
We can also check quickly with Anthony: @aaarendt Do you remember if there was another study before your 2002 paper doing "elevation change binning per elevation band" to compute glacier volume change (what we now call hypsometric gap-filling)?

adehecq commented on doc/source/dem_class.md:
```diff
@@ -81,9 +89,12 @@ import geoutils as gu
 # Plot the DEM and the vector file
 dem.plot(cmap="terrain", cbar_title="Elevation (m)")
-vect_gla.plot(dem)  # We pass the DEM as reference for the plot CRS/extent
+vect_gla.plot(dem, ec="k", fc="none")  # We pass the DEM as reference for the plot CRS/extent
```
Actually, on the figure, the plot extent is still that of the outlines, not the DEM...

rhugonnet replied:
Corrected, the reference is only for the CRS

adehecq commented on doc/source/dem_class.md:
```diff
@@ -140,7 +151,7 @@ dem_tba = xdem.DEM(filename_tba)
 dem_tba_coreg = dem_tba.coregister_3d(dem)
```
I was not aware of this method ! 😅

rhugonnet replied:
:wink:

adehecq commented on doc/source/dem_class.md:
````diff
@@ -157,10 +168,11 @@ stable terrain as a proxy.
 ```{code-cell} ipython3
 # Estimate elevation uncertainty assuming both DEMs have similar precision
-sig_dem, rho_sig = dem.estimate_uncertainty(dem_tba_coreg, precision_of_other="same")
+sig_dem, rho_sig = dem.estimate_uncertainty(dem_tba_coreg, precision_of_other="same", random_state=42)
````
I was also not aware of this method!!
Maybe explain in the comment why random_state is used and that it should not be used in the general case.

rhugonnet replied:
Done!

adehecq commented on doc/source/vertical_ref.md:
````diff
@@ -12,15 +12,64 @@ kernelspec:
-```{important}
+```{note}
 **A {class}`~xdem.DEM` already possesses a {class}`~xdem.DEM.crs` attribute that defines its 2- or 3D CRS**, inherited from
 {class}`~geoutils.Raster`. Unfortunately, most DEM products do not yet come with a 3D CRS in their raster metadata, and
````
Maybe it can be good also to explain that the 2D CRS may contain a term "datum" that is misleading. Even for 2D projections, one needs a datum, which is very often WGS84, for projecting from degrees to meters, but this is not the same as the vertical datum. There was some confusion about it at the winter school.

adehecq replied:
For example, projection UTM zone 44N (EPSG:32644), if printed as WKT string, starts with:
PROJCS["WGS 84 / UTM zone 44N",
    GEOGCS["WGS 84",
        **DATUM["WGS_1984",**
            SPHEROID["WGS 84",6378137,298.257223563,
                AUTHORITY["EPSG","7030"]],
            AUTHORITY["EPSG","6326"]],

rhugonnet replied:
A future improvement that would fit better on the guide pages, the feature pages should not get too much into details.

rhugonnet replied:
:wink:

adehecq commented on doc/source/terrain.md:
````diff
@@ -10,8 +24,87 @@ and tested for consistency against [gdaldem](https://gdal.org/programs/gdaldem.h
+slope = xdem.terrain.slope(dem.data, resolution=dem.res)
+```
+
+If computational performance is key, xDEM can rely on [RichDEM](https://richdem.readthedocs.io/) by specifying
````
Maybe add a note that we hope to speed-up calculation (using convoutions)?

adehecq replied:
Maybe you can also mention that if several attributes are needed, for optimization is is also better to call `get_terrain_attribute`? I know it is described later as well, but maybe have all "performance" related points at one location is good.

rhugonnet replied:
Thank god you commented here, I forgot to remove the RichDEM support since #558.
Rest done!

adehecq commented on doc/source/terrain.md:
```diff
@@ -10,8 +24,87 @@ and tested for consistency against [gdaldem](https://gdal.org/programs/gdaldem.h
+   :header-rows: 1
+   :stub-columns: 1
+
+   * - Attribute
```
Very useful table! Especially the unit column.
I am also thinking, should we raise a warning when trying to call many of these functions from geographic coordinates? Because it will return unrealistic values (e.g. slope).
Also for the hillshade, we could detect this and add the correction factor.

adehecq replied:
Added issue #634.

rhugonnet replied:
Perfect. I got exactly this issue the other day.

adehecq commented on doc/source/terrain.md:
````diff
@@ -20,20 +113,63 @@ The slope of a DEM describes the tilt, or gradient, of each pixel in relation to
-:width: 600
-```
+Finally, $\Delta x$
+and $\Delta y$ correspond to the pixel resolution west-east and south-north, respectively.
````
Add a note that we do not support non square pixels in xDEM at the moment? It will raise an error.

rhugonnet replied:
Added

adehecq commented on doc/source/terrain.md:
````diff
@@ -42,169 +178,212 @@ The aspect describes the orientation of strongest slope.
+with $p$ and $q$ defined in the slope section.
+
+```{warning}
+A north aspect represents the upper direction of the Y axis in the coordinate reference system of the
````
good point !!

adehecq commented on doc/source/terrain.md:
```diff
@@ -42,169 +178,212 @@ The aspect describes the orientation of strongest slope.
 
-The curvature map is the second derivative of elevation, which highlights the convexity or concavity of the terrain.
+The curvature is the second derivative of elevation, which highlights the convexity or concavity of the terrain.
 If a surface is convex (like a mountain peak), it will have positive curvature.
```
I thought we had discussed it before, but it's the opposite. convex = valley, concave = summit.

erikmannerfelt replied:
Uuuh what do you mean here @adehecq? A convex surface is summit-like while a concave surface is valley-like.

rhugonnet replied:
Thanks @erikmannerfelt, I also find several sources backing the current statements!
https://www.britannica.com/science/valley/Valley-cross-profiles
https://avalanche.ca/glossary/terms/convex-slope
Leaving as is :wink:

adehecq commented on doc/source/terrain.md:
```diff
@@ -42,169 +178,212 @@ The aspect describes the orientation of strongest slope.
 {func}`xdem.DEM.planform_curvature`
 
-The planform curvature is the curvature perpendicular to the direction of slope, reported in 100 m{sup}`-1`.
+The planform curvature is the curvature perpendicular to the direction of slope, reported in 100 m{sup}`-1`, also based
```
I had to check because I find the names counter-intuitive. To me, profile would mean perpendicular to the slope. I guess I am biased by the fact we often make cross profiles in glaciology 😉

rhugonnet replied:
Double-checked now too!

adehecq commented on doc/source/coregistration.md:
````diff
@@ -16,250 +16,472 @@ kernelspec:
-```{margin}
-<sup>1</sup>In a style resembling [scikit-learn's pipelines](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html#sklearn-linear-model-linearregression).
+```{tip}
+Often, an `inlier_mask` has to be passed to {func}`~xdem.coreg.Coreg.fit` to isolate static surfaces to utilize during coregistration (for instance removing vegetation, snow, glaciers). This mask can be easily derived using {func}`~geoutils.Vector.create_mask`.
````
I would recommend using the function `create_inlier_mask` from the workflow submodule.

rhugonnet replied:
This doesn't seem to have been limiting to users (and `workflows` is not in the API yet). Maybe we should discuss on the organization of this separately once you're back, also in relation to the CLI for coregistration.

adehecq commented on doc/source/coregistration.md:
```diff
@@ -16,250 +16,472 @@ kernelspec:
-The loop stops either when the maximum iteration limit is reached, or when the NMAD between the two products stops improving significantly.
+The [Nuth and Kääb (2011)](https://doi.org/10.5194/tc-5-271-2011) coregistration approach estimates a horizontal
+translation iteratively by solving a cosine equation between the terrain slope, aspect and the elevation differences.
+The iteration stops if it reaches the maximum number of iteration limit or if the tolerance does not improve.
```
it's not the tolerance that improves but the shift falls below a certain threshold.

rhugonnet replied:
Nice catch! :sweat_smile:

adehecq commented on doc/source/coregistration.md:
```diff
@@ -16,250 +16,472 @@ kernelspec:
 
-### Limitations
+Iterative Closest Point (ICP) coregistration is an iterative point cloud registration method from [Besl and McKay (1992)](https://doi.org/10.1117/12.57955). It aims at iteratively minimizing the distance between closest neighbours by applying sequential rigid transformations. If DEMs are used as inputs, they are converted to point clouds.
+As for Nuth and Kääb (2011), the iteration stops if it reaches the maximum number of iteration limit or if the tolerance does not improve.
```
same, tolerance does not improve

adehecq commented on doc/source/biascorr.md:
```diff
@@ -15,146 +17,420 @@ kernelspec:
+f, ax = plt.subplots(1, 2)
+ax[0].set_title("Before directional\nde-biasing")
+(tbc_dem_sumsin - ref_dem).plot(cmap='RdYlBu', vmin=-30, vmax=30, ax=ax[0])
+ax[1].set_title("After directional\nde-biasing")
```
The example plotted in the doc is not working... The bias is still very visible.

adehecq replied:
Actually, the bias seem to be doubled.

rhugonnet replied:
Yep, this is the ongoing issue described in the PR comment above that I can't seem to fix (it either corrects perfectly, or not at all... with the same `random_state`). :cry:

adehecq commented on doc/source/biascorr.md:
```diff
@@ -15,146 +17,420 @@ kernelspec:
-- **Recommended for:** Different native resolution between DEMs.
+- **Performs:** Correct biases along a terrain attribute.
+- **Supports weights:** Yes.
+- **Pros:** Useful to correct for instance curvature bias due to different native resolution between elevation data.
```
in Arnaud's project we ended up calling it the "resolution bias" as curvature is not the best proxy

rhugonnet replied:
Good point. Hard to change the name here as I add a bias based on curvature in the example below, we can adjust once we have the TPI implementation :wink:

adehecq commented on doc/source/biascorr.md:
```diff
@@ -15,146 +17,420 @@ kernelspec:
+- **Performs:** Correct biases along a terrain attribute.
+- **Supports weights:** Yes.
+- **Pros:** Useful to correct for instance curvature bias due to different native resolution between elevation data.
+- **Cons:** For curvature biases, only works for elevation data with relatively close natire resolution.
```
natire -> native

rhugonnet replied:
Nice catch

adehecq commented on doc/source/biascorr.md:
It looks like Blockwise coregistration is not documented in this section. Only in teh API and in the example. Would it make sense to have a dedicated section on "highly non linear bias correction"?

rhugonnet replied:
The section is already written but not added for now, it's copied in #584 in the meantime

rhugonnet replied:
Because of the errors still not fixed

adehecq commented on doc/source/gapfill.md:
```diff
@@ -0,0 +1,272 @@
+ddem = xdem.dDEM(raster=dem_2009 - dem_1990, start_time=dem_1990.datetime, end_time=dem_2009.datetime)
+
+# The example DEMs are void-free, so let's make some random voids.
+# Introduce 50000 nans randomly throughout the dDEM.
```
actually it is not 50000 nans, but 1/5 of all pixels.

adehecq commented on doc/source/gapfill.md:
```diff
@@ -0,0 +1,272 @@
+
+## Linear spatial interpolation
+
+Linear spatial interpolation (also often called bilinear interpolation) of dDEMs is arguably the simplest approach: voids are filled by an average of the surrounding pixels values, weighted by their distance to the void pixel.
```
Actually it is not a bilinear interpolation, but an Inverse Distance Weighting interpolation. Bilinear would be to interpolate between 4 pixels. At least according to wikipedia: "Bilinear interpolation considers the closest 2 × 2 neighborhood of known pixel values surrounding the unknown pixel's computed location."

rhugonnet replied:
That's an important detail! Changing everywhere.

adehecq commented on doc/source/gapfill.md:
```diff
@@ -0,0 +1,272 @@
+## Local hypsometric interpolation
+
+This approach assumes that there is a relationship between the elevation and the elevation change in the dDEM, which is often the case for glaciers.
+Elevation change gradients in late 1900s and 2000s on glaciers often have the signature of large melt in the lower parts, while the upper parts might be less negative, or even positive.
```
I would replace "melt" by thinning, as we cannot know if the thinning is due to melt or dynamic.

adehecq commented on doc/source/gapfill.md:
```diff
@@ -0,0 +1,272 @@
+# Plot before and after
+f, ax = plt.subplots(1, 2)
+ax[0].set_title("Before linear\ngap-filling")
+ddem.plot(cmap='RdYlBu', vmin=-10, vmax=10, ax=ax[0])
```
the color scale is very saturated. I would increase the vmin/vmax values.

adehecq commented on doc/source/gapfill.md:
````diff
@@ -0,0 +1,272 @@
+Then, voids are interpolated by replacing them with what "should be there" at that elevation, according to the model.
+
+```{code-cell} ipython3
+ddem_localhyps = ddem.interpolate(method="local_hypsometric", reference_elevation=dem_2009, mask=glaciers_1990)
````
The API for the function `interpolate` is not available, so it is difficult to get detailed info on the arguments. I was actually wondering about the type of argument `mask` as it looks like you are passing a Vector and not a mask.
I know this module needs more work, but good to raise an issue if this is not already mentioned.
I was also thinking that in the same way you generalized the bias correction to take any predictor variable, we could generalize the hypsometric method to work with any 1, 2 or Nd variables.

rhugonnet replied:
As it's glacier-focused, I think we need to discuss removing it altogether in its current state, including potentially the `dDEM` class. I've already added warning everywhere (API, pages and gallery examples) that these might be removed/re-factored.
Opening #636

adehecq commented on doc/source/gapfill.md:
```diff
@@ -0,0 +1,272 @@
+# Plot before and after
+f, ax = plt.subplots(1, 2)
+ax[0].set_title("Before regional\nhypsometric\ngap-filling")
+ddem.plot(cmap='RdYlBu', vmin=-10, vmax=10, ax=ax[0])
```
I could not find where the issue is, but on the figure "before gap-filling" there is no gap...

rhugonnet replied:
Fixed

adehecq commented on doc/source/uncertainty.md:
```diff
@@ -12,369 +14,385 @@ kernelspec:
+Frequently, in spatial statistics, a single correlation range is considered ("basic" method below).
+However, elevation data often contains errors with correlation ranges spanning different orders of magnitude.
+For this, [Rolstad et al. (2009)](http://dx.doi.org/10.3189/002214309789470950) and
+[Hugonnet et al. (2022)](http://dx.doi.org/10.1109/JSTARS.2022.3188922) considers
```
considers -> consider

adehecq commented on doc/source/uncertainty.md:
```diff
@@ -12,369 +14,385 @@ kernelspec:
+**Step 3: Derive final error**
 
-### Propagation of correlated errors
+And can now compute our final random error for the mean elevation change in this area of interest:
```
"and we can" -> "We can now"

adehecq commented on doc/source/code/intricacies_datatypes.py:
```diff
@@ -0,0 +1,58 @@
+ax[1, 0].set_yticklabels([])
+ax[1, 0].set_aspect("equal")
+
+plt.title("Types of elevation data")
```
it looks like this line is overwriting the "point cloud" title. Maybe you need to use "suptitle"?

rhugonnet replied:
Nice catch, yes!

adehecq commented on doc/source/elevation_intricacies.md:
```diff
@@ -0,0 +1,91 @@
+
+
+Additionally, there are a critical differences for elevation point clouds depending on point density:
+- **Sparse elevation point clouds** (e.g., altimetry) are generally be stored as small vector-type datasets (e.g., SHP). Due to their sparsity, for subsequent analysis, they are rarely gridded into a DEM, and instead compared with DEMs at the point cloud coordinates by interpolation of the DEM,
```
"be stored" -> "stored"

adehecq commented on doc/source/robust_estimators.md:
```diff
@@ -1,17 +1,21 @@
-- Robust measures for the central tendency (e.g., mean) and dispersion (e.g., standard deviation), to evaluate DEM quality and converge during {ref}`coregistration`,
-- Robust measures for estimating spatial autocorrelation for uncertainty analysis in {ref}`spatialstats`.
+In order to mitigate their effect, the analysis of elevation data can integrate [robust statistics](https://en.wikipedia.org/wiki/Robust_statistics) at different levels:
+- **Robust estimators for the central tendency (e.g., mean) and dispersion (e.g., standard deviation)**, to evaluate DEM quality and converge during {ref}`coregistration`,
```
do you mean "median" and "NMAD" instead of mean and std?

adehecq replied:
"converge" -> "convergence"? I am not sure if you mean "evaluate convergence" or something else

rhugonnet replied:
It's true it was a bit convoluted, reworded!

adehecq commented on doc/source/robust_estimators.md:
pretty cool !

rhugonnet commented on doc/source/gapfill.md:
```diff
@@ -0,0 +1,272 @@
+
+Similarly to local hypsometric interpolation, the elevation change is assumed to be largely elevation-dependent.
+With the regional approach (often also called "global"), elevation change gradients are estimated for all glaciers in an entire region, instead of estimating one by one.
+This is advantageous in respect to areas where voids are frequent, as not even a single dDEM value has to exist on a glacier in order to reconstruct it.
```
Removed

rhugonnet commented:
Accounted for all comments, opened issues otherwise. All tests passing.
Merging!

vschaffn commented:
@rhugonnet It's probably not deliberate, but the logging.md doc has been removed from your updates, and the logging_configuration.py example has been renamed plot_logging_configuration.py

rhugonnet commented:
In short: None of the two pages were rendering after #610. For instance, renaming into "plot_xxx" is required for it to be rendered by Sphinx-Gallery, otherwise it is shown without execution of the code.
In the comment above I explain how you can check the documentation rendering in the future with your own RTD fork setup :slightly_smiling_face:

vschaffn commented:
@rhugonnet Sorry, I had forgotten about that comment, thanks for the reminder 😃
